source/agora/network/NetworkClient.d

```diff
@@ -208,18 +176,58 @@ class NetworkClient
 
             If the request failed, returns an empty array
 
+        Throws:
+            Exception if the request failed.
+
     ***************************************************************************/
 
     public Block[] getBlocksFrom (ulong block_height, size_t max_blocks)
     {
-        try
-        {
-            return this.api.getBlocksFrom(block_height, max_blocks);
-        }
-        catch (Exception ex)
+        return this.attemptRequest(
+            this.api.getBlocksFrom(block_height, max_blocks), this.exception);
+    }
+
+    /***************************************************************************
+
+        Attempt a request up to 'this.max_retries' attempts, and make the task
+        wait this.retry_delay between each attempt.
+
+        If all requests fail and 'ex' is not null, throw the exception.
+
+        Params:
+            log_level = the logging level to use for logging failed requests
+            T = the type of delegate
+            dg = the delegate to call
+            ex = the exception to throw if all attempts fail (if not null)
+
+        Returns:
+            the return value of dg(), which may be void
+
+    ***************************************************************************/
+
+    private T attemptRequest (LogLevel log_level = LogLevel.error, T)(
```

Comment:
Does log level has to be a template parameter ?

Reply:
Answered IRL before, but yes because `log` takes it as a CT param, while `doDebug` which takes it as a runtime parameter is private.